In [227]:
import os
import glob
import pandas as pd                
import numpy as np
import matplotlib.pyplot as plt

In [299]:
df = pd.read_csv('orders.csv')
df1 = pd.read_csv('order_items.csv')
df.head(-3)

id       ordered_at     delivered_at delivered_to
0        0  8/19/2018 22:01  8/19/2018 22:19       CX-767
1        1  11/16/2018 0:12  11/16/2018 0:40       CX-847
2        2  11/27/2018 8:06  11/27/2018 8:08       CX-972
3        3  10/14/2018 1:16  10/14/2018 1:25       CX-390
4        4  11/28/2018 5:10  11/28/2018 5:16       CX-914
...    ...              ...              ...          ...
4991  4992  11/21/2018 3:44  11/21/2018 4:09       CX-900
4992  4993  10/30/2018 9:32  10/30/2018 9:58        CX-80
4993  4994   11/7/2018 0:05   11/7/2018 0:32       CX-470
4994  4995  11/28/2018 2:58  11/28/2018 3:12       CX-778
4995  4996  11/1/2018 19:46  11/1/2018 20:16       CX-337

[4996 rows x 4 columns]

In [229]:
df1['order_id'].nunique() ## 4909 unique order ids
df1

item_id  order_id               food_item  amount_paid
0            0       114    Braised Beef Noodles        10.25
1            1      2825             Gam Ja Tang        11.25
2            2      4717             Cha Siu Bao         9.25
3            3      1035  Black Sugar Bubble Tea         5.25
4            4      1023          Pork Belly Bao         9.25
...        ...       ...                     ...          ...
19995    19995      3561  Passionfruit Green Tea         3.75
19996    19996      3246             Gam Ja Tang        11.25
19997    19997       604             Gam Ja Tang        11.25
19998    19998      4130      Matcha Cheese Cake         6.75
19999    19999      3626  Black Sugar Bubble Tea         5.25

[20000 rows x 4 columns]

In [230]:
def meal_type(df1):
    
    if(df1['food_item'] == 'Black Sugar Bubble Tea'):
        return 'Dessert'
    elif (df1['food_item'] == 'Braised Beef Noodles'):
        return 'Dinner'
    elif (df1['food_item'] == 'Cha Siu Bao'):
        return 'Snacks'
    elif (df1['food_item'] == 'Gam Ja Tang'):
        return 'Dinner'
    elif (df1['food_item'] == 'Herbal Grass Jelly Drink'):
        return 'Beverages'
    elif (df1['food_item'] == 'Matcha Cheese Cake'):
        return 'Dessert'
    elif (df1['food_item'] == 'Passionfruit Green Tea'):
        return 'Beverages'
    elif (df1['food_item'] == 'Pork Belly Bao'):
        return 'Snacks'
    elif (df1['food_item'] == 'Red Bean & Taro Slush'):
        return 'Beverages'

        
df1['meal_type'] = df1.apply(meal_type, axis = 1)

In [231]:
df1

item_id  order_id               food_item  amount_paid  meal_type
0            0       114    Braised Beef Noodles        10.25     Dinner
1            1      2825             Gam Ja Tang        11.25     Dinner
2            2      4717             Cha Siu Bao         9.25     Snacks
3            3      1035  Black Sugar Bubble Tea         5.25    Dessert
4            4      1023          Pork Belly Bao         9.25     Snacks
...        ...       ...                     ...          ...        ...
19995    19995      3561  Passionfruit Green Tea         3.75  Beverages
19996    19996      3246             Gam Ja Tang        11.25     Dinner
19997    19997       604             Gam Ja Tang        11.25     Dinner
19998    19998      4130      Matcha Cheese Cake         6.75    Dessert
19999    19999      3626  Black Sugar Bubble Tea         5.25    Dessert

[20000 rows x 5 columns]

In [232]:
df_order_id = df1.groupby(['order_id'])["food_item"].count()
df_order_id
mealtype =  df1.groupby(['order_id'])["meal_type"].value_counts()
mealtype

order_id  meal_type
0         Snacks       2
          Beverages    1
1         Beverages    4
          Dinner       2
          Snacks       1
                      ..
4997      Beverages    1
          Dinner       1
          Snacks       1
4998      Beverages    1
          Dinner       1
Name: meal_type, Length: 11441, dtype: int64

In [233]:
id_mealtype = df1.groupby(['order_id'])['meal_type'].value_counts().unstack().fillna('')
#id_mealtype['Beverages'] = id_mealtype['Beverages'].apply(int)
#id_mealtype['Dessert'] = id_mealtype['Dessert'].apply(int)
#id_mealtype['Dinner'] = id_mealtype['Dinner'].apply(int)
#id_mealtype['Snacks'] = id_mealtype['Snacks'].apply(int)

id_mealtype.dtypes

meal_type
Beverages    object
Dessert      object
Dinner       object
Snacks       object
dtype: object

In [234]:
id_mealtype.to_csv('mealtype.csv')
df2 = pd.read_csv('mealtype.csv')
df2['order_id'].nunique() ##4909 unique Order ids

4909

# mealtype order by each order id

In [235]:
df2

order_id  Beverages  Dessert  Dinner  Snacks
0            0        1.0      NaN     NaN     2.0
1            1        4.0      NaN     2.0     1.0
2            2        NaN      NaN     2.0     NaN
3            3        1.0      NaN     1.0     2.0
4            4        NaN      NaN     2.0     1.0
...        ...        ...      ...     ...     ...
4904      4994        1.0      1.0     1.0     NaN
4905      4995        NaN      NaN     6.0     1.0
4906      4996        2.0      NaN     2.0     1.0
4907      4997        1.0      NaN     1.0     1.0
4908      4998        1.0      NaN     1.0     NaN

[4909 rows x 5 columns]

In [236]:
df['id'].nunique() ## 4999 umqiue ids

4999

In [300]:
from dateutil.parser import parse
#df['Time_diff'] =   df['delivered_at'] - df['ordered_at']
  
df['delivered_at'] = pd.to_datetime(df['delivered_at'], format='%m/%d/%Y %H:%M')
df['ordered_at'] = pd.to_datetime(df['ordered_at'], format='%m/%d/%Y %H:%M')
#df['Time_diff'] =   df['delivered_at'] - df['ordered_at']

## difference of time
DIFF_Time_spend = df['delivered_at']  - df['ordered_at']
##DIFF_Time_spend  ## total time difference
df['Time_spend_mins'] = DIFF_Time_spend.dt.components.minutes
df

id          ordered_at        delivered_at delivered_to  \
0        0 2018-08-19 22:01:00 2018-08-19 22:19:00       CX-767   
1        1 2018-11-16 00:12:00 2018-11-16 00:40:00       CX-847   
2        2 2018-11-27 08:06:00 2018-11-27 08:08:00       CX-972   
3        3 2018-10-14 01:16:00 2018-10-14 01:25:00       CX-390   
4        4 2018-11-28 05:10:00 2018-11-28 05:16:00       CX-914   
...    ...                 ...                 ...          ...   
4994  4995 2018-11-28 02:58:00 2018-11-28 03:12:00       CX-778   
4995  4996 2018-11-01 19:46:00 2018-11-01 20:16:00       CX-337   
4996  4997 2018-11-13 01:51:00 2018-11-13 02:21:00       CX-537   
4997  4998 2018-10-16 20:04:00 2018-10-16 20:17:00        CX-98   
4998  4999 2018-09-16 02:45:00 2018-09-16 03:09:00       CX-203   

      Time_spend_mins  
0                  18  
1                  28  
2                   2  
3                   9  
4                   6  
...               ...  
4994               14  
4995               30  
4996               30  
4997               13  
4998               24  

[4999 rows x 5 columns]

In [310]:
deliver_sort =  df.groupby(['delivered_to'])['id'].count().sort_values()
df.delivered_to.nunique() ## 992 unqiue delivery sites
df.sort_values(by=['delivered_to'])
df_refined = df[['id','delivered_to','Time_spend_mins']]
df_refined

id delivered_to  Time_spend_mins
0        0       CX-767               18
1        1       CX-847               28
2        2       CX-972                2
3        3       CX-390                9
4        4       CX-914                6
...    ...          ...              ...
4994  4995       CX-778               14
4995  4996       CX-337               30
4996  4997       CX-537               30
4997  4998        CX-98               13
4998  4999       CX-203               24

[4999 rows x 3 columns]

In [318]:
df_refined.groupby('delivered_to')['id'].count().sort_values()

delivered_to
CX-748     1
CX-91      1
CX-866     1
CX-625     1
CX-141     1
          ..
CX-6      23
CX0       23
CX-5      25
CX-9      28
CX-4      32
Name: id, Length: 992, dtype: int64

In [320]:
df_refined['Delivery_freq'] = df_refined.groupby('delivered_to')['id'].transform('count')
#df_refined
df_refined.sort_values(by=['Delivery_freq'])
df_refined

<ipython-input-320-70dbd3d27750>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_refined['Delivery_freq'] = df_refined.groupby('delivered_to')['id'].transform('count')


id delivered_to  Time_spend_mins  Delivery_freq
0        0       CX-767               18              5
1        1       CX-847               28              7
2        2       CX-972                2              6
3        3       CX-390                9              7
4        4       CX-914                6              4
...    ...          ...              ...            ...
4994  4995       CX-778               14              8
4995  4996       CX-337               30              6
4996  4997       CX-537               30              2
4997  4998        CX-98               13              8
4998  4999       CX-203               24              8

[4999 rows x 4 columns]

In [329]:
df2 = df2.rename(columns={'order_id':'id'})
df2

id  Beverages  Dessert  Dinner  Snacks
0        0        1.0      NaN     NaN     2.0
1        1        4.0      NaN     2.0     1.0
2        2        NaN      NaN     2.0     NaN
3        3        1.0      NaN     1.0     2.0
4        4        NaN      NaN     2.0     1.0
...    ...        ...      ...     ...     ...
4904  4994        1.0      1.0     1.0     NaN
4905  4995        NaN      NaN     6.0     1.0
4906  4996        2.0      NaN     2.0     1.0
4907  4997        1.0      NaN     1.0     1.0
4908  4998        1.0      NaN     1.0     NaN

[4909 rows x 5 columns]

In [347]:
df3 = pd.merge(df_refined, df2, on=['id'], how='inner')
df4 = df3.sort_values(by=['Delivery_freq'])
df4.columns

Index(['id', 'delivered_to', 'Time_spend_mins', 'Delivery_freq', 'Beverages',
       'Dessert', 'Dinner', 'Snacks'],
      dtype='object')

In [334]:
df4.dtypes

id                   int64
delivered_to        object
Time_spend_mins      int64
Delivery_freq        int64
Beverages          float64
Dessert            float64
Dinner             float64
Snacks             float64
dtype: object

In [353]:
df4 = df4.fillna('0')
df4['Beverages'] = df4['Beverages'].apply(np.int64)
df4['Dessert'] = df4['Dessert'].apply(np.int64)
df4['Dinner'] = df4['Dinner'].apply(np.int64)
df4['Snacks'] = df4['Snacks'].apply(np.int64)

In [371]:
df4.dtypes
df4['Customer_status'] = df4['Delivery_freq'].apply(lambda x: 'New' if x == 1 else 'old')

In [372]:
df4.to_csv('refined.csv', index=False)

In [373]:
df5 = pd.read_csv('refined.csv')

In [374]:
df5

id delivered_to  Time_spend_mins  Delivery_freq  Beverages  Dessert  \
0     3910        CX-75               11              1          1        0   
1     3704       CX-321               10              1          0        0   
2      955       CX-363               25              1          0        0   
3     3706       CX-251               19              1          1        1   
4     2399       CX-510               13              1          0        0   
...    ...          ...              ...            ...        ...      ...   
4903  2311         CX-4               15             32          2        1   
4904  2450         CX-4                4             32          2        0   
4905   851         CX-4               11             32          1        0   
4906  2914         CX-4               18             32          2        0   
4907  1510         CX-4               19             32          1        0   

      Dinner  Snacks Customer_status  
0          1       1             New  
1          3       1             New  
2          1       3             New  
3          0       1             New  
4          1       1             New  
...      ...     ...             ...  
4903       1       3             old  
4904       1       1             old  
4905       1       3             old  
4906       2       2             old  
4907       2       3             old  

[4908 rows x 9 columns]

In [394]:
old_new_status = df5['Customer_status'].value_counts()

In [395]:
old_new_status

old    4862
New      46
Name: Customer_status, dtype: int64

# old customers - 4862 and New customers - 46

# when the delivery_frequency is 1, it is a new address/new customer
# when the delivey_frequency is more than 1, it is an existing customer